implement auth0 guard

Tamrael opened the pull request:
### Changes

- new guard driver "auth0" to allow for easy authentication without custom middleware and allow for testing with actingAs
- bind Auth0UserRepository to it's contract on plugin boot to allow for customization from inside laravel apps while using the guard
- guard is inactive as long as the config values are not set to use this guard as driver in laravel auth config. See readme for example on how to set them

### References

Implementation through [closure request guard](https://laravel.com/docs/6.x/authentication#closure-request-guards)

Implemented to fix testing problems mentioned in #161 

### Testing

- [x] This change has been tested on the latest version Laravel

### Checklist

- [x] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [x] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)

Tamrael commented:
@joshcanhelp let me try and address your questions to the best of my knowledge
- the guard only replaces the Middleware as the point of authorization for requests with a jwt bearer token. Login/logout still need to be handled as well as any Userprovider changes and such
- I would suggest replacing the part about the middleware with guidance of the guard because it's a lot closer to the laravel way of doing authentication. For the sake of people that already implemented middleware I would leave it in as footnote or alternative way of handling API authentication.
- the guard helper this is using was added in 5.6 and the API never changed, only concerns is the "bearerToken" function that's not well documented. Can't find when this was added atm but I'm almost sure it's a helper that existed a long time.

joshcanhelp commented:
@Tamrael - Thanks again! 

Let me get this running locally and if everything looks good, we'll get this in.

Tamrael commented:
@joshcanhelp just had the time to check. the bearerToken function is part of the "InteractsWithInput" trait and is available since 5.4 so everything about this guard is 5.4 and above. if you need anything else I'd be glad to help.

Tamrael commented on README.md:
````diff
@@ -46,6 +46,36 @@ You can implement your own cache strategy by creating a new class that implement
+```
+'guards' => [
+    ...
+    'auth0' => [
````
the name of the guard can be anything. only the driver is important. the provider has to be the name of the provider defined below

Tamrael commented on README.md:
```diff
@@ -46,6 +46,36 @@ You can implement your own cache strategy by creating a new class that implement
+
+'providers' => [
+    ...
+    'auth0' => [
```
the name of the provider can be anything as long as the driver is correctly set

joshcanhelp commented:
@Tamrael - Apologies for the delay here. 

I appreciate all your explanation here and can see that this would be useful to add. I pulled the branch down and followed the documentation you added. Everything came together as expected but when I hit the route, I get a redirect to the Auth0 login page. If I understand how this works, it's looking for a JWT in the headers with `$request->bearerToken()`. Since a user would not explicitly add that to the headers their browsers send, I would assume that this guard is used on API-type routes where a 302 to a login page would be difficult for the API caller to handle. I would have expected this to fail on the route with a 401 or similar. 

Am I misunderstanding how this should be used? I followed your setup instructions with the following in `routes/api.php`:

```php
Route::middleware('auth:auth0')->get('/auth0-guard', function (Request $request) {
    return $request->user();
});
```

Tamrael commented:
@joshcanhelp sorry didn't think to mention that. the request to the api must expect json as a return or be ajax (not pjax) and accept any content type. the handling for the unauthenticated message is completely done by laravel.
see [Exceptions/Handler.php#L216](https://github.com/laravel/framework/blob/d4edbf5f2569c066d6e931cacb3a61d96bdfa313/src/Illuminate/Foundation/Exceptions/Handler.php#L216) for the implementation

joshcanhelp commented:
@Tamrael - Thank you for the suggestion, that worked as expected. 

I'm sorry to keep coming back with questions/changes but I think we should get this in. The next version, however, is a major so this should be merged into that branch. When I switch the branch here to the dev branch `7.0.0-dev`, I get a few conflicts. Would you mind switching the base branch and making those changes? 

Thank you once again for sticking with this! I'm glad we'll be able to get this in.

Tamrael commented:
@joshcanhelp i rebased this onto 7.0. tests ran in laravel 5, 6 and 7. can't tests this further atm but I'm pretty sure this works.

I removed the tests/bootstrap.php because it's not needed when setting the ini in the phpunit.xml, hope you don't mind.
Had to fix the function declarations for the setUp of the tests because phpunit > 7 was required for laravel 7 compatibility

joshcanhelp commented on README.md:
```diff
@@ -46,6 +46,36 @@ You can implement your own cache strategy by creating a new class that implement
 
+### Using auth guard
+
+To use the auth0 guard add it to the config/auth.php with the driver "auth0":
```
```suggestion
To use the `auth0` guard, add it to `config/auth.php` with the driver `auth0`:
```

joshcanhelp commented on README.md:
````diff
@@ -46,6 +46,36 @@ You can implement your own cache strategy by creating a new class that implement
+],
+```
+
+After that simply use just like any other guard e.g.:
````
```suggestion
Once that has been added, add the guard to the middleware of any API route:
```

joshcanhelp commented:
@Tamrael - We appreciate your work on this! Thanks for the great communication as well!
